### ISSUE TITLE
Avoid copying Transferable objects

### ISSUE THREAD

**Issue by ssube** [audreyt/node-webworker-threads#58]
I'm attempting to use the library to run a few background threads in a node app, each processing an 8-32MB Uint8Array. Running with 4 workers and 4MB of data each, it currently takes more than 15 minutes to post the initial messages (before it can even *start* processing). The performance is rather bad, since typed arrays are supposed to be Transferable.

So far as I can tell, each array is being serialized repeatedly when being transferred between threads. This appears to be happening [here](https://github.com/audreyt/node-webworker-threads/blob/master/src/worker.js#L34) (thanks to [apsillers for finding that](http://chat.stackoverflow.com/transcript/message/24192748#24192748)). Producing those strings peaks around 7GB of RES memory.

My test app consists of two files:

 - index.js:
```
var Worker = require('webworker-threads').Worker;

var workerCount = 4, finishedCount = 0, bufferSize = (1024 * 1024 * 4);
var workers = {};
var buffers = {};

function messageReceiver(event) {
  var data = event.data;
  var id = data.id;
  var buffer = data.buffer;
  console.log('Received', buffer[0], 'from', id);
  workers[id].terminate();
  if (++finishedCount === workerCount) {
    console.log('All workers finished.');
  }
}

for (var i = 0; i < workerCount; ++i) {
  console.log('Spawning worker', i);
  var worker = new Worker('worker.js');
  worker.onmessage = messageReceiver;
  workers[i] = worker;
}
console.log('All workers spawned.');

for (var i = 0; i < workerCount; ++i) {
  console.log('Allocating buffer', i);
  buffers[i] = new Uint32Array(new Buffer(bufferSize));
}
console.log('All buffers allocated.');

for (var i = 0; i < workerCount; ++i) {
  console.log('Launching worker', i);
  var w = workers[i];
  var a = buffers[i];
  var data = {
    id: i,
    buffer: a.buffer
  };
  w.postMessage(data, [data.buffer]);
}
console.log('All workers launched.');
```

 - worker.js:
```
function finishWorker(id, buffer) {
  console.log('Thread ' + id + ' finished filling buffer.');
  var data = {
    id: id,
    buffer: buffer
  };
  postMessage(data, [data.buffer]);
}

function fillBuffer(buffer) {
  for (var i = 0; i < buffer.byteLength; ++i) {
    if (i % (1024 * 1024) === 0) {
      console.log('Thread ' + id + ' filled ' + i + ' bytes');
    }
    buffer[i] = ((Math.random() * 255) | 0) % 255;
  }
}

this.onmessage = function (event) {
  var id = event.data.id, buffer = event.data.buffer;
  console.log('Thread ' + id + ' received buffer of ' + buffer.byteLength);
  fillBuffer(buffer);
  finishWorker(id, buffer);
};
```

**Comment by audreyt** [comment]
Yes, that would be highly desirable!

I did look at Chromium codebase to see how they handled Transferables, but my C++ fu is less than stellar and couldn't quite figure it out. Help welcome!

**Comment by mlogan** [comment]
I'm going to work on this - the basic approach is call `v8::ArrayBuffer::Externalize` to release the creating thread's claim to the memory, and then create new ArrayBuffers and new Arrays in the receiving thread with the existing memory. (I'm not sure how chromium does it, but I believe we'll have to also create weak PersistantHandles for the newly created ArrayBuffers in order to free the memory upon garbage collection).

**Comment by audreyt** [comment]
Awesome, thanks @mlogan ! :rainbow:

**Comment by mlogan** [comment]
What's the correct way to run the tests? `npm test` does nothing. Should I just be doing `for f in test/*; do node $f; done`?

**Comment by mlogan** [comment]
FYI, I won't be able to get this working in node 0.10. That version of v8 doesn't expose the necessary APIs. 0.12 and iojs should be fine though.

**Comment by audreyt** [comment]
Yes it's all manual. Making `npm test` harness (with mocha or jest or something) the tests would be great (perhaps in another issue).

That is fine — we have a long-standing issue with 0.12+ documented in https://github.com/audreyt/node-webworker-threads/issues/48#issuecomment-76603020 — if that's taken care of, I have no problem to upgrade the dep to 0.12+.

**Comment by mlogan** [comment]
Ok, noob question: I've gotten most of the C++ written, but now I need to make some changes to `src/worker.js`. It appears that that file gets compiled to `src/worker.js.c` for use in the native module. How do I update `src/worker.js.c` from `src/worker.js`?

**Comment by mlogan** [comment]
And of course, 10 seconds after I send my previous message, I find the answer: `npm run js`

**Comment by mlogan** [comment]
Ok, I have most of the C++ stuff done, but there's a few thorny issues that are preventing me from finishing this. See the #59 for detailed discussion.
